### PR TITLE
Store password hashes and fix Login API to use them

### PR DESCRIPTION
**Pull request by Goro7777**
The hashing algorithm used is PBKDF2. It uses SHA512 internally.

The arguments of the PBKDF2 call (the salt, the number of iterations, the hashing algorithm used internally, and the key size) are appended to the hash generated by PBKDF2 and stored as part of the password hash in users.json.

### CODE REVIEW

File: Technolab.OnlineLibrary.Web/Controllers/AuthController.cs [hunk 1]
```diff
@@ -60,13 +62,50 @@ private async Task PerformLogin(User user)
...
+            var identity = new ClaimsIdentity(claims, CookieAuthenticationDefaults.AuthenticationScheme);
+
             await HttpContext.SignInAsync(
                 CookieAuthenticationDefaults.AuthenticationScheme,
                 new ClaimsPrincipal(identity));
         }
 
         private ILibraryDbContextFactory ContextFactory { get; }
+
+        private bool VerifyPassword(string enteredPassword, string correctHashWithParams)
```
Comment (sarustamyan) [inline Technolab.OnlineLibrary.Web/Controllers/AuthController.cs]: It's probably worth moving both this and the `GenerateHash` methods to the `User` class itself

File: Technolab.OnlineLibrary.Web/Controllers/AuthController.cs [hunk 2]
```diff
@@ -60,13 +62,50 @@ private async Task PerformLogin(User user)
...
+            HashAlgorithmName hashAlgorithm = new HashAlgorithmName(split[3]);
+            int keySize = Convert.ToInt32(split[4]);
+
+            var enteredPasswordHash = Rfc2898DeriveBytes.Pbkdf2(enteredPassword, salt, iterations, hashAlgorithm, keySize);
+            return CryptographicOperations.FixedTimeEquals(enteredPasswordHash, Convert.FromHexString(correctHash));
+        }
+
+        private string GenerateHash(string password)
+        {
+            int keySize = 32;
```
Comment (sarustamyan) [inline Technolab.OnlineLibrary.Web/Controllers/AuthController.cs]: Generally speaking avoid magic string/numbers. If anything, these settings should come from the config file.

File: Technolab.OnlineLibrary.Web/Controllers/AuthController.cs [hunk 3]
```diff
@@ -60,13 +62,50 @@ private async Task PerformLogin(User user)
...
+            HashAlgorithmName hashAlgorithm = HashAlgorithmName.SHA512;
+
+            var hash = Rfc2898DeriveBytes.Pbkdf2(
+                Encoding.UTF8.GetBytes(password),
+                salt,
+                iterations,
+                hashAlgorithm,
+                keySize);
+
+            string hashStr = Convert.ToHexString(hash) + ":" +
```
Comment (sarustamyan) [inline Technolab.OnlineLibrary.Web/Controllers/AuthController.cs]: Move the `":"` into a `const` and use it both here and in `VerifyPassword`

**Review by sarustamyan — CHANGES_REQUESTED** [review]

File: Technolab.OnlineLibrary.Web/Controllers/AuthController.cs (hunk 2 above)
Comment (Goro7777) [reply Technolab.OnlineLibrary.Web/Controllers/AuthController.cs]: You mean the Consts.cs file, right? I'll put the delimiter (":") there as well in that case, if there are no objections.

**Review by sarustamyan — APPROVED** [review]

**Pull request merged**